### DISCUSSION TITLE
With a read only filesystem, which temporary folders must be made writeable?

### DISCUSSION THREAD

**Discussion by erikschul** [keycloak/keycloak#36142] [Q&A] [answered]
In Kubernetes, a security practice is to make the image read only, and add writable temporary folders (emptyDir).

Which temporary (empty) folders must be made writable for Keycloak to work, beyond its data directory?

It fails on start because it tries to build, but the filesystem is read only.

**Comment by ora7598** [comment, marked as answer]
I would go with the following emptyDir s:
- /tmp - under root path
- /opt/writable-keycloak - which will contain a copy of the entire original KC folder

Instead of creating specific writable folders (which can be problematic in the future in case of file structure changes), I went on raising an init-container that copies the content from `/opt/keycloak` to the new emptyDir I specified above `/opt/writable-keycloak`.
And then you gonna need your entrypoint to be something like: `./opt/writable-keycloak/bin/kc.sh start`

**Comment by shawkins** [comment]
This has also been discussed on #11286 and several other issues.
